src/dune_engine/fs_memo.ml

```diff
@@ -2,10 +2,47 @@ open! Stdune
 open! Import
 open Memo.Build.O
 
+type t = { dune_file_watcher : Dune_file_watcher.t option }
+
+(* Ideally this should be an [Fdecl], but there are currently two reasons why
+   it's not:
+
+   - We read the workspace file before we start the inotify watcher, so [init]
+   is called after [t_ref] is used. This is clearly not good (it means we never
+   re-read the workspace file)
+
+   - There are tests that call [Scheduler.go] multiple times, therefore [init]
+   gets called multiple times. Since they don't use file watcher, it shouldn't
+   be a problem. *)
+let t_ref = ref { dune_file_watcher = None }
+
+let init ~dune_file_watcher =
+  match !t_ref.dune_file_watcher with
+  | Some _ ->
+    Code_error.raise
+      "Called [Fs_memo.init] a second time after a file watcher was already \
+       set up "
+      []
+  | None -> t_ref := { dune_file_watcher }
+
 (* Files and directories have non-overlapping sets of paths, so we can track
    them using the same memoization table. *)
 let memo =
-  Memo.create "fs_memo" ~input:(module Path) (fun _path -> Memo.Build.return ())
+  Memo.create "fs_memo"
+    ~input:(module Path)
+    (fun path ->
```

Comment:
As discussed offline, it feels weird that we are going to call `add_watch path` on every invalidation of the cell. It seems that this isn't a problem because `add_watch` is idempotent.

Interestingly, if `path` disappears, we do have to call `add_watch` on its parent, which suggests that maybe it's not weird to have this logic here after all.

@aalekseyev Shall we add (a version of) the above text as a comment?

Reply:
Added a comment like this:

```
      (* It may seem weird that we are adding a watch on every invalidation of
         the cell. This is OK because [add_watch] is idempotent, in the sense
         that we are not accumulating watches.

         In fact, if path disappears then we lose the watch and have to
         re-establish it, so doing it on every computation is sometimes
         necessary. *)
```

by the way, I changed the code somewhat when making the fix to the "watch before init" issue, so you may want to re-read some of it.